### ISSUE TITLE
Upgrade Alpine image to v18

### ISSUE THREAD

**Issue by kalbfled** [department-of-veterans-affairs/notification-api#1295]
## User Story - Business Need
We need to upgrade to the latest version of Alpine 3.18 because there are a number of longstanding security issues with our current version, 3.16, and 3.16's EOL for security fixes is 23 May 2024.

### User Story(ies)
**As a** VA Notify stakeholder,
**I want** the application image built upon Alpine Linux v3.18 or later
**So that** the application minimizes security vulnerabilities.

- [ ]Ticket is understood, and QA has been contacted (if the ticket has a QA label).

### Additional Info and Resources
There is a new version of urllib3 (v2.0.0) that can't be installed in projects that use botocore because, in the setup.py file, it has the version restriction urllib3>=1.25.4,<1.27. See the current requirement [here](https://github.com/boto/botocore/blob/develop/setup.py#L28).  Also see the [associated ticket](https://github.com/boto/botocore/issues/2926) in the botocore repository.

That same issue contains a workaround [in this comment](https://github.com/boto/botocore/issues/2926#issuecomment-1673963526) to utilize the latest version of botocore - author is the maintainer for awscli/boto3:
> You can currently install urllib3 2.0.x with boto3/botocore/awscli and it will work as we've unblocked Gentoo and nixpkgs to do. The primary reason that the pin is still in place for general python installations is because urllib3 2.0 has compatibility issues with Amazon Linux 2 and some versions of AWS Lambda. These are detailed in the urllib3 migration guide [here](https://urllib3.readthedocs.io/en/stable/v2-migration-guide.html#ssl-module-is-compiled-with-openssl-1-0-2-k-fips).
> 
> Because of the way pip works by always choosing to install the latest version of packages, moving the pin for urllib3 will unexpectedly break everyone not pinning urllib3 or using OpenSSL 1.1.1+. That's something we take very seriously with AWS products and try to ensure never happens. We're still working on a plan to smoothly migrate this customer base, but don't have any immediate timeline.
> 
> For the time being, users are free to install urllib3 2.x _after_ they've installed boto3/botocore/awscli, and should be able to use each package. Something like this would work in automated environments returning a successful response code. We realize it's not ideal, but is the most tractable solution we have currently that maintains stability:
> 
> ```
> python -m pip install boto3==x.y.z
> python -m pip install urllib3==2.0.4 --no-deps
> ```

**NOTE** at the time of this writing, urllib3 should be pinned to 2.0.6. When picking up this ticket, be sure to check what the latest 2.x.x for urlib3 is. See also the comment added 10/6 regarding a botocore PR recently merged that resolves this issue for python 3.10+.

The notification-api image currently [builds from Alpine 3.16](https://github.com/department-of-veterans-affairs/notification-api/blob/master/ci/Dockerfile#L6).  Multiple tickets require upgrading to a newer version of Alpine. See the dependencies section of this ticket.

Upgrading Alpine to v18 to get OpenSSL 3.1.1--is distinct from the task of upgrading the Python "requests" package. The requirements are as follows:

-     botocore requires urllib3>=1.25.4,<1.27 
-     requests 2.31.0 requires urllib3>=1.21.1,<3 
-     urllib3 >= 2.0.0 requires openssl >= 1.1.1 
-     Alpine 16 (currently in use) provides openssl 1.1.1; Alpine 18 provides openssl 3.1.1

Neither "requests" nor "urllib3" are top level requirements for the Flask app, although 3 lambda functions require "requests".  #1241 (closed) upgraded requests to v2.31.0, and "urllib3" to v1.26.x

We need separate tickets to upgrade the "requests" lambda layer used by the lambda functions.

## Engineering Checklist

- [x] Upgrade the "FROM" base image in Dockerfile and Dockerfile.local to use Alpine >=3.18
- [x] Update the table [here](https://github.com/department-of-veterans-affairs/notification-api#maintaining-docker-images)
- [x] Implement the workaround for installing boto3 and urllib3
- [x] Build the image locally

## Acceptance Criteria

- [x] The app runs locally and that all unit tests pass
- [x] All QA regression tests pass

## QA Considerations
- [ ] beat this thing up in regression testing
- [ ] Twilio callbacks verified to work without error

## Potential Dependencies
_Leave blank if n/a_

## Out of Scope
_Leave blank if n/a_

**Comment by kalbfled** [comment]
@k-macmillan @mjones-oddball @cris-oddball Upon further study, I've concluded that this task--upgrading Alpine to v18 to get OpenSSL 3.1.1--is distinct from the task of upgrading the Python "requests" package.  The requirements are as follows:

- botocore requires `urllib3>=1.25.4,<1.27` ([source](https://github.com/boto/botocore/blob/develop/setup.py#L28))
- requests 2.31.0 requires `urllib3>=1.21.1,<3` ([source](https://github.com/psf/requests/blob/main/setup.py#L64))
- urllib3 >= 2.0.0 requires openssl >= 1.1.1 ([source](https://github.com/urllib3/urllib3/blob/main/CHANGES.rst#200-2023-04-26))
- Alpine 16 (currently in use) provides openssl 1.1.1; Alpine 18 provides openssl 3.1.1

Neither "requests" nor "urllib3" are top level requirements for the Flask app, although 3 lambda functions require "requests".  I expect that "requests" will be upgraded to v2.31.0, and "urllib3" to v1.26.x, as part of #1241, which is in progress.  We need separate tickets to upgrade the "requests" lambda layer used by the lambda functions.

I propose that we keep this ticket as-is for upgrading Alpine.

**Comment by mjones-oddball** [comment]
@kalbfled are you making the requests lambda layer tickets or is Baldwin aware and creating those?

**Comment by cris-oddball** [comment]
@kalbfled while we are awaiting botocore to support urllib3 >= 2.x.x, can we at least upgrade to Alpine 3.17 instead of jumping to 3.18?

**Comment by kalbfled** [comment]
@mjones-oddball I created an API and an infra ticket.  See #1329.

@cris-oddball Negative.  [Alpine 3.17 includes OpenSSL v3.0.9](https://pkgs.alpinelinux.org/packages?name=openssl&branch=v3.17&repo=&arch=x86_64&maintainer=), and we can't upgrade to any OpenSSL version above 1.1.1 without upgrading urllib3.

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/1295?planning-poker) with Zenhub @cris-oddball @EvanParish @justaskdavidb2 @k-macmillan @kalbfled @ldraney @nikolai-efimov

**Comment by k-macmillan** [comment]
@cris-oddball You mentioned you have a workaround documented. Please include it here when you are able.

**Comment by cris-oddball** [comment]
@k-macmillan turns out I had added the workaround directly into the description. :)

**Comment by cris-oddball** [comment]
@k-macmillan Note that this fix didn't work for everyone. On 10/6/2023, botocore merged a [PR](https://github.com/boto/botocore/pull/3034) that allows urllib3 for python 3.10, so at the end of the day, if the above hack doesn't work, then we should re-focus on upgrading the api to use python 3.10 to get us to alpine 3.18.

**Comment by pquentin** [comment]
> and we can't upgrade to any OpenSSL version above 1.1.1 without upgrading urllib3.

Hello 👋 urllib3 maintainer here. Sorry for jumping in your issue, but urllib3 1.26.x does support OpenSSL 1.1.1 and 3.

**Comment by cris-oddball** [comment]
Testing Twistlock on this ticket to see which vulnerabilities were cleared.

**Comment by cris-oddball** [comment]
Cleared over 20 vulnerabilities in twistlock and closed 8 tickets. There are still two remaining vulnerabilities for which I will open new Twistlock tickets.

Diverting testing for Twilio callbacks verified to work without error to the python upgrade ticket and closing this one.
